// jsnative/index.ts



import { TestBed, TestModuleMetadata, async } from '@angular/core/testing';

import { } from '@angular/core';

export function jasmine_describe( desc: string, fn: Function ): void {
    describe( desc, () => { fn.call( null ); } );
}

export function jasmine_it( desc: string, fn: Function ): void {
    it( desc, () => { fn.call( null ); } );
}


export function jasmine_expect<T>( obj: T ): jasmine.Matchers<T> {
    return expect( obj );
}

export function jasmine_beforeEach( action: ( p1: DoneFn ) => void ) {
    return beforeEach( action );
}

// angular

export class TestModuleMetadataBuilder {
    private data: TestModuleMetadata = {};

    public imports( ...classes: any[] ): TestModuleMetadataBuilder {
        this.data.imports = classes;
        return this;
    }
    public providers( ...classes: any[] ): TestModuleMetadataBuilder {
        this.data.providers = classes;
        return this;
    }
    public declarations( ...classes: any[] ): TestModuleMetadataBuilder {
        this.data.declarations = classes;
        return this;
    }

    public build(): TestModuleMetadata {
        return this.data;
    }
}

export function angular_async( fn: Function ): Function {
    return async( fn );
}

export function angular_newTestModuleMetadataBuilder(): TestModuleMetadataBuilder {
    return new TestModuleMetadataBuilder();
}
